docs/architecture/adr-024-decentralized-identifiers-registry.md

```diff
@@ -0,0 +1,396 @@
+    
+    // External controller in DID syntax who can manage the verification method (optional).
+    // Documentation: https://w3id.org/security#controller
+    string controller = 3;
+
+    // Material of the public key in the verification method.
+    // Documentation: https://www.w3.org/TR/did-core/#verification-material
+    oneof verification_material {
+        // JSON web key format public key in JSON.
+        string public_key_jwk = 4;
+        
+        // Base58 encoded public key.
+        string public_key_multibase = 5;
+    };
+}
+
+// Service represents the ways of communicating with DID subjects or associated entities.
+// Documentation: https://www.w3.org/TR/did-core/#services
+message Service {
+    // Unique identifier for the service.
+    // Example: did:desmos:<unique-identifier>#<service-name>
+    string id = 1;
+
+    // Type of the service.
+    string type = 2;
+
+    // Endpoints of service.
+    repeated string service_endpoint = 3;
+}
+```
+
+#### DidMetadata
+
+`DidMetadata` represents the metadata to show the latest status of DID document.
+
+The store keys of `DIDMetadata` for KVStore will be:
+
+* 0x2 | {id} -> ProtocolBuffer(DidMetadata)
+
+##### Protobuf definition
+
+```protobuf
+// DidMetadata represents metadata of the DID document.
+// Documentation: https://www.w3.org/TR/did-core/#did-document-metadata
+message DidMetadata {
+    // The unique identifier in DID syntax of DID document.
+    string id = 1;
+
+    // Timestamp of the DID document creation.
+    google.protobuf.Timestamp created = 2;
+
+    // Timestamp of the last DID document update.
+    google.protobuf.Timestamp updated = 3;
+
+    // Flag that indicates DID document is deactivated or not.
+    bool deactivated = 4;
+
+    // Last version id of the DID document.
+    uint64 version_id = 5;
+}
+```
+
+### `Msg` Service
+
+We will allow the following operations to be performed:
+
+- Create a new DID document
+- Update an existing DID document
+- Deactivate DID document
+- Edit verification inside DID document
+
+> NOTE  
+> When updating DID document, sender's public key MUST be its controller or one of keys inside capability invocation field.
+
+#### Protobuf definition
+
+```protobuf
+service Msg{
+    // CreateDidDoc allows to create a newly DID document.
+    rpc CreateDidDoc(MsgCreateDidDoc) returns(MsgCreateDidDocResponse);
+
+    // UpdateDidDoc allows to update an existing DID document.
+    rpc UpdateDidDoc(MsgUpdateDidDoc) returns(MsgUpdateDidDocResponse);
+
+    // DeactivateDidDoc allows to deactivate DID document.
+    rpc DeactivateDidDoc(MsgsDeactivateDidDoc) returns(MsgsDeactivateDidDocResponse);
+
+    // EditVerificationMethod allows to edit an existing verification method inside DID document.
+    rpc EditVerificationMethod(MsgEditVerificationMethod) returns(MsgEditVerificationMethodResponse);
+}
+
+// MsgCreateDidDoc represents the message to be used to create a new DID document.
+message MsgCreateDidDoc {
```

Comment:
I see that inside this message there are some fields that contain some data that can ideally be faked by the user. For example I have the following questions:

1. How can we make sure that the data provided by the user is correct (e.g. the verification methods are all valid)? 
2. Should we allow the user to set the ID? Wouldn't it be better to generate this one instead? 
3. Should we allow the user to set the Controller ID?

Reply:
> How can we make sure that the data provided by the user is correct (e.g. the verification methods are all valid)?
We can only check the public keys is the valid format or not. On the other hand, if user provide the non-owned public key to these fields, then user can not provide the proper signature to verifier, it means that we can recognize that user does not have that DID document.

> Should we allow the user to set the ID? Wouldn't it be better to generate this one instead?
Yeah, you are right. Generating on chain would be better for this case. I will update the docs.

> Should we allow the user to set the Controller ID?
Controller is a spec that allowing other DID to update the document. It gives more flexibility to DID document control, like group control. To have the flexibility, I think we can have it.
https://www.w3.org/TR/did-core/#the-relationship-between-did-controllers-and-did-subjects